src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/Logging/NativeAotFatalErrorException.cs

```diff
@@ -0,0 +1,45 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System;
+
+using ILCompiler.Logging;
+
+namespace ILCompiler
+{
+    /// <summary>
+    /// Represents a known error that occurred during compilation time which is not solvable by the user.
+    /// This is used when we want to present the non-recoverable error with a specific error code.
+    /// </summary>
+    public class NativeAotFatalErrorException : Exception
+    {
+        public MessageContainer MessageContainer { get; }
+
+        /// <param name="message">Error message with a description of what went wrong</param>
+        public NativeAotFatalErrorException(MessageContainer message)
+            : base(message.ToString())
+        {
+            if (message.Category != MessageCategory.Error)
+                throw new ArgumentException($"'{nameof(NativeAotFatalErrorException)}' ought to be used for errors only");
+
+            if (message.Code == null || message.Code.Value == 0)
+                throw new ArgumentException($"'{nameof(NativeAotFatalErrorException)}' must have a code that indicates a failure");
```

Comment:
These should be asserts. There's only few things that are worse than hitting a rare error codepath and getting an ArgumentException that loses all the context because the rare error path was untested.